x-pack/plugins/fleet/server/services/agent_policy.ts

```diff
@@ -1312,14 +1315,15 @@ class AgentPolicyService {
     updatedPolicies: Array<Partial<AgentPolicy>> | null;
     failedPolicies: Array<{ id: string; error: Error | SavedObjectError }>;
   }> {
-    const { saved_objects: agentPoliciesWithEnabledAgentTamperProtection } =
-      await soClient.find<AgentPolicySOAttributes>({
-        type: SAVED_OBJECT_TYPE,
-        page: 1,
-        perPage: SO_SEARCH_LIMIT,
-        filter: normalizeKuery(SAVED_OBJECT_TYPE, 'ingest-agent-policies.is_protected: true'),
-        fields: ['revision'],
-      });
+    const agentPolicyFetcher = this.fetchAllAgentPolicies(soClient, {
+      kuery: 'ingest-agent-policies.is_protected: true',
+    });
+
+    const agentPoliciesWithEnabledAgentTamperProtection = [];
+
+    for await (const agentPolicy of agentPolicyFetcher) {
+      agentPoliciesWithEnabledAgentTamperProtection.push(...agentPolicy);
```

Comment:
Do you need to store all data in memory?

This would be more efficient if each page of data could be processed (bulk updated) from here before moving on to the next page data. That would also avoid having to store all of the agent policies in memory (which could be allot of data). Doing that will also improve the bulk update in that it will never have more than 1000 items in the payload. Also note that in the past we implemented limits around how many items to place in the bulk update to also make that even more efficient (I think for artifacts).

Reply:
Great idea, thanks!
Refactored.